task/thermal/src/control.rs

```diff
@@ -116,84 +142,216 @@ impl InputChannel {
     post_failed_count: u32,
 }
 
-impl<'a, B: BspT> ThermalControl<'a, B> {
+/// Represents a temperature reading at the time at which it was taken
+#[derive(Copy, Clone, Debug)]
+enum TemperatureReading {
+    /// Normal reading, timestamped using monotonic system time
+    Valid { time_ms: u64, value: Celsius },
+
+    /// This sensor is not used in the current power state
+    Inactive,
+}
+
+/// Configuration for a PID controller
+pub struct PidConfig {
+    pub gain_p: f32,
+    pub gain_i: f32,
+    pub gain_d: f32,
+}
+
+struct PidState {
+    /// Previous (time, input) tuple, for derivative term
+    prev_error: Option<f32>,
+
+    /// Accumulated integral term
+    integral: f32,
+}
+
+impl PidState {
+    /// Attempts to drive the error to zero.
+    ///
+    /// The error and output are expected to have the same signs, i.e. a large
+    /// positive error will produce a large positive output.
+    fn run(
+        &mut self,
+        cfg: &PidConfig,
+        error: f32,
+        limit: core::ops::Range<f32>,
+    ) -> f32 {
+        let p_contribution = cfg.gain_p * error;
+
+        // Only accumulate the integral term if we're not saturated, to prevent
+        // integral wind-up.
+        if limit.contains(&p_contribution) {
+            self.integral += error;
+        }
+        let i_contribution = cfg.gain_i * self.integral;
+
+        let d_contribution = if let Some(prev_error) = self.prev_error {
+            (error - prev_error) * cfg.gain_d
+        } else {
+            0.0
+        };
+        self.prev_error = Some(error);
+
+        let out = p_contribution + i_contribution + d_contribution;
+        if out > limit.end {
+            limit.end
+        } else if out < limit.start {
+            limit.start
+        } else {
+            out
+        }
+    }
+}
+
+impl Default for PidState {
+    fn default() -> Self {
+        Self {
+            prev_error: None,
+            integral: 0.0,
+        }
+    }
+}
+
+/// This corresponds to states shown in RFD 276
+enum ThermalControlState {
+    /// Wait for each sensor to report in at least once
+    Boot {
+        values: [Option<TemperatureReading>; bsp::NUM_TEMPERATURE_INPUTS],
+    },
+
+    /// Normal happy control loop
+    Running {
+        values: [TemperatureReading; bsp::NUM_TEMPERATURE_INPUTS],
+        pid: PidState,
+    },
+
+    /// In the overheated state, one or more components has entered their
+    /// critical temperature ranges.  We turn on fans at high power and record
+    /// the time at which we entered this state; at a certain point, we will
+    /// timeout and drop into `Uncontrolled` if components do not recover.
+    Overheated {
+        values: [TemperatureReading; bsp::NUM_TEMPERATURE_INPUTS],
+        start_time: u64,
+    },
+
+    /// The system cannot control the temperature; power down and wait for
+    /// intervention from higher up the stack.
+    Uncontrollable,
+}
+
+enum ControlResult {
+    Pwm(PWMDuty),
+    PowerDown,
```

Comment:
A question from earlier, will we ever need an intermediate state "CoolDown" if we're overheated and are shutting down?

Reply:
I don't think so.  The `Overheating` state is meant to be emergency cooling; `PowerDown` is if all else has failed and it cuts power immediately.